klab.hub/src/main/java/org/integratedmodelling/klab/hub/api/EngineAuthResponeFactory.java

```diff
@@ -91,16 +96,49 @@ public EngineAuthenticationResponse getRespone(EngineAuthenticationRequest reque
 		return null;
 	}
 	
+    private void validateProfileAccountStatus(ProfileResource profile) throws AuthenticationFailedException {
+        switch(profile.accountStatus) {
+        case active:
+            return;
+        case locked:
+            throw new AuthenticationFailedException(String.format("User '%s' is locked.", profile.getUsername()));
+        case deleted:
+            throw new AuthenticationFailedException(String.format("User '%s' is deleted.", profile.getUsername()));
+        case pendingActivation:
+        case verified:
+            throw new AuthenticationFailedException(String.format("User '%s' has not completed the activation process.", profile.getUsername()));
+        case expired:
+        default:
+            throw new AuthenticationFailedException(String.format("User '%s' is not active.", profile.getUsername()));
+        }
+    }
+
 	@SuppressWarnings("unchecked")
     private EngineAuthenticationResponse remoteEngine(ProfileResource profile,
-			String cipher, LicenseConfiguration config) throws NoSuchProviderException, IOException, PGPException {
+			String cipher, LicenseConfiguration config) throws NoSuchProviderException, IOException, PGPException, AuthenticationFailedException {
 		Properties engineProperties = PropertiesFactory.fromProfile(profile, config).getProperties();
 		Properties cipherProperties = new CipherProperties().getCipherProperties(config, cipher);
 		ArrayList<HubNotificationMessage> messages = new ArrayList<HubNotificationMessage>();
-		
-		if(profile.accountStatus == AccountStatus.locked) {
-			throw new LockedUserException(profile.getUsername());
-		}
+
+        validateProfileAccountStatus(profile);
+
+        // TODO for now, we just assume that there is a single agreement per profile.
+        List<Agreement> validAgreements = profile.getAgreements().stream()
+                .filter(a -> !a.isRevoked() && !a.isExpired()).collect(Collectors.toList());
```

Comment:
as commented, we could move the logic `!a.isRevoked() && !a.isExpired()` to a `a.isValid()` method